src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -591,3 +646,288 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+ * parts of the tree multiple times. When we walk a node which contains multiple
+ * sub-trees we recurse and remember which parts of the tree contained a Var. If some
+ * did and some didn't we evaluate the parts of the tree which didn't. If none of the
+ * sub-trees contained a Var we do nothing besides reporting that fact.
+ *
+ * At the top-level of the expression we evaluate everything if no Vars were found.
+ */
+static Node * PartiallyEvaluateExpression(Node *expression)
+{
+	ExpressionWalkerState unused;
+	InitializeWalkerState(&unused);
+	unused.isTopLevel = true;
+
+	return PartiallyEvaluateExpressionWalker(expression, &unused);
+}
+
+
+static Node * PartiallyEvaluateExpressionWalker(Node *expression,
+		ExpressionWalkerState *parentState)
+{
+	ExpressionWalkerState childState;
+	Node *copy = NULL;
+	bool isBranch = false;
+	bool containsVar = false;
+
+	if (expression == NULL)
+	{
+		return expression;
+	}
+
+	/* pass any argument lists back to the mutator to copy and recurse for us */
+	if (IsA(expression, List))
+	{
+		return expression_tree_mutator(expression,
+									   PartiallyEvaluateExpressionWalker,
+									   parentState);
+	}
+
+	InitializeWalkerState(&childState);
+
+	if (IsA(expression, Var))
+	{
+		containsVar = true;
+	}
+
+	if (IsA(expression, FuncExpr) || IsA(expression, OpExpr))
+	{
+		/* Tell the next step that it should add itself to one of the subtree lists */
+		childState.parentIsBranch = true;
+		isBranch = true;
+	}
+
+	copy = expression_tree_mutator(expression,
+								   PartiallyEvaluateExpressionWalker,
+								   &childState);
+
+	containsVar |= childState.containsVar;
+
+	if (parentState->parentIsBranch)
+	{
+		if (containsVar)
+		{
+			parentState->subtreesWithVar = lappend(parentState->subtreesWithVar, copy);
+			parentState->subtreeOrder = lappend_int(parentState->subtreeOrder, 0);
+		}
+		else
+		{
+			parentState->subtreesWithoutVar = lappend(parentState->subtreesWithoutVar,
+													  copy);
+			parentState->subtreeOrder = lappend_int(parentState->subtreeOrder, 1);
+		}
+	}
+
+	/* never set containsVar=false because parentState is shared between sibling trees */
+	if (containsVar)
+	{
+		parentState->containsVar = true;
+	}
+
+	if (isBranch && parentState->isTopLevel &&
+		list_length(childState.subtreesWithVar) == 0)
+	{
+		return EvaluateExpression(expression);
+	}
+
+	if ((list_length(childState.subtreesWithVar) != 0) &&
+		(list_length(childState.subtreesWithoutVar) != 0))
+	{
+		if (IsA(expression, FuncExpr))
+		{
+			((FuncExpr *) copy)->args = RewriteArgs(&childState);
+		}
+
+		if (IsA(expression, OpExpr))
+		{
+			((OpExpr *) copy)->args = RewriteArgs(&childState);
+		}
+	}
+
+	return copy;
+}
+
+/* evaluate every non-Var arg and return an argument list */
+static List * RewriteArgs(ExpressionWalkerState *state)
+{
+	List *rewrittenArgs = NULL;
+	ListCell *nextArgCell = NULL;
+	int nextArg = 0;
```

Comment:
A variable name like nextFromVarlist or something would make this easier to understand.